### ISSUE TITLE
Ideal redux API

### ISSUE THREAD

**Issue by vslinko** [gaearon/redux#48]
I have experiment about ideal redux API for me: https://github.com/vslinko/observable-components-test

1. State should be a real observable. So I choice Rx as best observable implementation.

2. `wrapActionCreator(actionCreator)()` replaced by `performAction(actionCreator())`.

3. Pass `performAction()` to async action instead of `dispatch()`. Thats simplifies async action API: https://github.com/vslinko/observable-components-test/blob/master/src/actions/items.js#L25

4. Flux shouldn't be dependent on React.

What are you think?

**Comment by ooflorent** [comment]
> 1\. State should be a real observable. So I choice Rx as best observable implementation.

I think this should belong to userland. Redux should perform independently of the state implementation.

> 2\. `wrapActionCreator(actionCreator)()` replaced by `performAction(actionCreator())`
> 3\. Pass `performAction()` to async action instead of `dispatch()`.

I prefer the current implementation using `dispatch`. Maybe I'm missing something.

> 4\. Flux shouldn't be dependent on React.

That's also my opinion. Flux implementation should be as low-level as possible but decoupling the impl from React would also make React components or decorators more complex.

**Comment by gaearon** [comment]
I'll take a better look at your repo a bit later but I don't want a hard observable dependency. I want to make it easy to plug one in though.

**Comment by vslinko** [comment]
> I think this should belong to userland. Redux should perform independently of the state implementation.
>
> I'll take a better look at your repo a bit later but I don't want a hard observable dependency. I want to make it easy to plug one in though.
You are right. User can wrap `createRedux` by `createRxRedux`.

> I prefer the current implementation using dispatch. Maybe I'm missing something.
`dispatch` cat perform only sync actions but `performAction` can perform sync actions and async actions. So, when you want call one async action from another you should do this:
```js
function oneAsyncAction() {
  return (dispatch, state) => {
    // you should know anotherAsyncAction is async and pass dispatch and state even your async action doesn't use state
    anotherAsyncAction()(dispatch, state)
  }
}

function oneAsyncAction() {
  return (performAction) => {
    // you're don't care about is anotherAction sync or async and about it's arguments
    performAction(anotherAction())
  }
}
```

**Comment by vslinko** [comment]
> You are right. User can wrap createRedux by createRxRedux.

https://github.com/vslinko/observable-components-test/commit/0a5119f0ffdd2c7c2fd62998aa508f29e7350695

**Comment by gaearon** [comment]
>4. Flux shouldn't be dependent on React.

Yep, I agree here. After #46 is done, I'll extract dispatcher again, with a real public API for it.

**Comment by gaearon** [comment]
Good point about `performAction`. Can you find a way to combine this neatly with approach I'm suggesting [here](https://github.com/gaearon/redux/pull/46#issuecomment-109580095?

**Comment by gaearon** [comment]
@vslinko I extracted `Dispatcher`: 6954066fb999320847e885c178a008eabf3edd09

**Comment by vslinko** [comment]
I'm defined minimal flux API https://gist.github.com/vslinko/cab24085f029def8997b

**Comment by vslinko** [comment]
> I extracted Dispatcher

Great, after you extract `createDispatcher` `redux` will me isomorphic

**Comment by vslinko** [comment]
Ow, I did not noticed `Dispatcher` is class, not component.

**Comment by vslinko** [comment]
Closed because `0.8.0`
